packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -2,22 +2,1869 @@
+class _BorderContainer extends StatefulWidget {
+  const _BorderContainer({
+    Key key,
+    @required this.border,
+    @required this.gap,
+    this.child
+  }) : assert(border != null),
+       assert(gap != null),
+       super(key: key);
+
+  final _InputBorder border;
+  final _InputBorderGap gap;
+  final Widget child;
+
+  @override
+  _BorderContainerState createState() => new _BorderContainerState();
+}
+
+class _BorderContainerState extends State<_BorderContainer> with SingleTickerProviderStateMixin {
+  AnimationController _controller;
+  Animation<double> _animation;
+  _InputBorderTween _border;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = new AnimationController(
+      duration: _kTransitionDuration,
+      vsync: this,
+    );
+    _animation = new CurvedAnimation(
+      parent: _controller,
+      curve: _kTransitionCurve,
+    );
+    _border = new _InputBorderTween(
+      begin: widget.border,
+      end: widget.border,
+    );
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  void didUpdateWidget(_BorderContainer oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.border != oldWidget.border) {
+      _border = new _InputBorderTween(
+        begin: oldWidget.border,
+        end: widget.border,
+      );
+      _controller
+        ..value = 0.0
+        ..forward();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new CustomPaint(
+      foregroundPainter: new _InputBorderPainter(
+        repaint: new Listenable.merge(<Listenable>[_animation, widget.gap]),
+        animation: _animation,
+        border: _border,
+        gap: widget.gap,
+        textDirection: Directionality.of(context),
+      ),
+      child: widget.child,
+    );
+  }
+}
+
+// Paints an [InputDecorator]'s outline or underline border.
+//
+// The border separates the input decorator's "container", i.e. the
+// optionally filled area above the optional helper, error, and counter
+// fields and adjacent to the optional icon.
+//
+// When `borderType` is [InputBorderType.underline] a horizontal
+// line is drawn at the bottom of the container. The line's color
+// and weight are defined by the `borderSide` parameter.
+//
+// When `borderType` is [InputBorderType.outline] a rounded rectangle
+// outline is drawn. The line's color and weight are defined by the
+// `borderSide` parameter, and the (circular) radii of the corners are
+// defined by the `borderRadius` parameter. The outline border may
+// have an animated gap, see [paint].
+class _InputBorder extends ShapeBorder {
+  _InputBorder({
+    this.borderType: InputBorderType.underline,
+    this.borderSide: BorderSide.none,
+    this.borderRadius: BorderRadius.zero,
+    this.gapPad: 4.0,
+    this.gapAnimation,
+  }) : assert(borderSide != null),
+       assert(borderRadius != null),
+       assert(_cornersAreCircular(borderRadius)),
```

Comment:
can you elaborate on why it's important that they be circular?

Reply:
The label text's gap can extend into the corners (even both the top left and top right corners) and I didn't want to bother with the more complicated problem of finding how far the gap penetrated into an elliptical corner. It didn't seem to be an important feature to support now. I suppose the summary rationale is: sloth.